### ISSUE TITLE
TURN configuration

### ISSUE THREAD

**Issue by christophschw** [jsxc/jsxc#402]
[ejabberd.yml.txt](https://github.com/jsxc/jsxc/files/590701/ejabberd.yml.txt)

Hey,
I got troubles with the TURN server configuration. I registered an account at numb.viagenie.ca. Video is working if clients are in the same network, but not if there are in different. The known problem when dialling:

> ICE_connection_failure
> Anruf beendet: jingle_reason_failed-transport.
> Anruf beendet: jingle_reason_failed-transport.

Configuration: owncloud 9.0.4
XMPP: ejabbered
Server NGINX
OS: CentOS 7 with firewalld (I already stopped it temporarily)

> XMPP domain: server's IP
> BOSH URL: /http-bind/ 

→ is reachable (because of SLL redirect the proxy pass directs to https://localhost/http-bind)

> TURN URL: numb.viagenie.ca
> TURN Username: as registered, my email address 
> TURN Password: as registered
> TURN Secret: empty
> TURN TTL: 86400

Which information should I provide? How can I debug?
Thanks in advance,
Christoph

**Comment by sualko** [comment]
Did you try to logout and login again? It is hard to debug this kind of issue. In Chrome there is chrome://webrtc-internals/ but I also had to use wireshark in the past. Which browser do you use?

**Comment by christophschw** [comment]
I tested with Chromium 54 and Firefox 49 and logged out/in after each try. No luck.

Sorry for the file extension mix: Couldn't upload an .ods...
[output_webrtc_internals.ods.xlsx](https://github.com/jsxc/jsxc/files/592824/output_webrtc_internals.ods.xlsx)

Thank you!

**Comment by dteleguin** [comment]
Just FYI - I was testing similar setup last days, and what I've found is that WebRTC features (file transfer and voice/video) simply do not work under latest Firefox versions, including 49, 50 and dev build. Of the four Firefox × Chrome combinations, only Chrome-Chrome pair has proven to be working. So if you intend to isolate STUN/TURN-specific problems, I'd recommend using Chrome-Chrome first of all.

@sualko do I file a separate issue for that?

**Comment by christophschw** [comment]
Thanks for the information @dteleguin ! But it works with FF 49 – if they clients are in the same network... So I guess it is a real TURN problem.

**Comment by dteleguin** [comment]
Sorry for my somewhat premature statements, I've managed to sort it out more or less.

First, voice/video connection didn't work because there was no camera at the Firefox box. Without camera attached, it simply fails with "media failure: NotFoundError" upon getUserMedia(). Could we allow voice-only connections? Most corporate desktops today don't have cameras AFAIK.

Second, file transfer didn't work because of buggy Firefox extension that blocked WebRTC functionality. Hint to all Firefox users: if you encounter file transfer and/or voice/video problems with JSXC, first of all try running Firefox with a clean profile or disable extensions installed. After I had nuked that extension, I finally managed to send files from Firefox to Chrome.

The inverse, however, didn't work. When sending from any browser to Firefox, the latter seems to receive nothing at all. Neither there is an incoming message, nor a new WebRTC session is established. @christophschw, could you please test if it works for you?

**Comment by dteleguin** [comment]
@christophschw could you also try the [Trickle ICE](https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/) test both under Chrome and Firefox? Remove the Google server and add yours (don't forget the URL should be "turn:numb.viagenie.ca"), then press "Gather candidates".

What do you get? Ideally, there should be a list of your local IP(s) (as "host"), NAT external IP (as "srflx") plus TURN server IP (as "relay"). I've figured out that Firefox gets the complete list, while Chrome receives everything but the relay IP. That happens because Chrome fails to send authentication data to TURN server, despite the corresponding fields having been filled.

If both peers are under symmetric NAT, TURN is the only option. So this might be exactly your case.

**Comment by christophschw** [comment]
This is the ouput:

Time	Component	Type	Foundation 	Protocol	Address	Port	Priority
0.001	1	host	0	UDP	192.168.1.221	52706	126 | 32512 | 255
0.001	2	host	0	UDP	192.168.1.221	57978	126 | 32512 | 254
0.265	Not reachable?

With another account on same machine:
0.002	1	host	0	UDP	192.168.1.221	40157	126 | 32512 | 255
0.003	2	host	0	UDP	192.168.1.221	55382	126 | 32512 | 254

When I add "turn:" in my JSXC settings calls don't even pop up at the other machine... 

@sualko – What is the right syntax for the TURN server? 
Thanks guys

**Comment by christophschw** [comment]
First time working! Thanks for the hint @dteleguin – Chromium to Mobile FF to Chromium in other network is working

But why no FF?

**Comment by dteleguin** [comment]
@christophschw congratulations on the finally working configuration :) as for Firefox, I have suspicion that it's ICE stack is somehow incompatible with Numb's software (particularly with the combination of attributes it returns with TURN Allocate Successful message). Not sure if it's a Firefox or Numb bug however. Do you have a machine to deploy your own TURN server for testing purposes? I can help you setup [coturn](https://github.com/coturn/coturn) daemon if you wish.

**Comment by christophschw** [comment]
I'm not sure how to test it but I guess this missing feature could be the reason for the lack of TURN support in FF:
https://bugzilla.mozilla.org/show_bug.cgi?id=1056934
What do you think? Did you encounter a working FF environment that you tested with TrickleICE?

@dteleguin Thank you for the offer! This would be very nice but unfortunately I don't own an additional machine – except of my Raspberry II with dyndns in my network... ;)

**Comment by sualko** [comment]
Hey guys. I am sorry that you have trouble with WebRTC. In the past it was most often an issue with the turn server, but WebRTC is no draft standard yet, so it could be always the case, that things broke in a new browser version. I try to update the corresponding libraries for 3.1. to get the latest goodies from them. For voice-only calls see this open issue #6.

**Comment by popindavibe** [comment]
You'll need TURN only for specific LAN router's settings (Symetrical NAT).
By default JSXC will aim for google's public STUN server, and that will be enough for you peers to set up the connection and get it working most of the time.
Your configuration should also reflect that and try STUN, then TURN if STUN fails. 
I'll create a wiki part on it, as I eventually got it working with **coturn** to act as STUN and TURN server. Tested successfully with Firefox 45 and 49 versions, chromium and Firefox on Android.

**Comment by dteleguin** [comment]
@popindavibe cool! That wiki page is needed so much, thank you for your effort! BTW, could you please share a working coturn config? I've only managed to make it work by enabling the `server-relay` option, which is marked as "non-standard and dangerous". Without it, clients get stuck with STUN (that doesn't work due to symmetric NATs) and don't even try switch to TURN mode. I'm using latest stable coturn (4.5.0.4).

To everybody: Google's STUN server is hardcoded into JSXC default config; troubleshooting WebRTC issues should always start with looking up `stun.l.google.com` in DNS. I clearly remember there was a period when it resolved to 127.0.0.1. Not sure though if it was Google's fault or my provider's.

**Comment by popindavibe** [comment]
@dteleguin Thanks! I had to post-pone the second part, and will probably get back to the first part, because I realized my coturn config in JSXC is visible from outside, being in the javascript, and the javascript being loaded right at the the login page of SOGo. 
@sualko  Did I miss something here or is it the design that implies my turn creds are visible even for non-connected users? I started looking at ephemeral credentials as a may-be solution, but didn't find time nor courage to punch it through. 

This is my turnserver.conf 
```
listening-port=3478
tls-listening-port=5349
alt-listening-port=3479
alt-tls-listening-port=5350

listening-ip=PUB.IP.N.1
listening-ip=PUB.IP.N.2
relay-ip=PUB.IP.N.1

min-port=49152
max-port=65535
verbose

lt-cred-mech
server-name=example.com
userdb=/var/lib/turn/turndb
realm=example.com
cert=/etc/tls/certs/coturn_ca.crt
pkey=/etc/tls/private/coturn.key
dh-file=/etc/turn/dhparam.pem
no-stdout-log
log-file=/var/log/turn/turn.log

no-sslv3
no-tlsv1
```
That's about it, didn't try to dig further, I was actually surprised how coturn did the job pretty fast.

And then, in sjsxc.config.js (for SOGo), at the same level than _xmpp_ array 
```
RTCPeerConfig: {
            /** Time-to-live for config from url */
            ttl: 3600,

            /** [optional] If set, jsxc requests and uses RTCPeerConfig from this url */
            url: null,
            /** If true, jsxc send cookies when requesting RTCPeerConfig from the url above */
            withCredentials: false,

            /** ICE servers like defined in http://www.w3.org/TR/webrtc/#idl-def-RTCIceServer */
            iceServers: [{
               urls: 'stun:stun.example.com'
               },
               {
               urls: ['turn:turn.example1.com','turn:turn.example2.com'],
               username: 'turnUser',
               credential: 'XXXXXXXXXXXX',
               credentialType: 'password'
            }]
        }
```

**Comment by dteleguin** [comment]
@popindavibe,

> By default JSXC will aim for google's public STUN server, and that will be enough for you peers to set up the connection and get it working most of the time.

I'm afraid [that's not the case](https://github.com/jsxc/jsxc/blob/master/src/jsxc.lib.options.js#L258):

```
    iceServers: [{
        urls: 'stun:stun.stunprotocol.org'
    }]
```
and... _(drumroll)_
```
[user@localhost ~]$ host stun.stunprotocol.org
stun.stunprotocol.org has address 127.0.0.1
```
I've made a misleading comment earlier - it's not Google's STUN server that resolves to 127.0.0.1, it's stun.stunprotocol.org. And AFAIK, it doesn't have any relation to Google. @sualko, seems like JSXC ships with non-working STUN config out-of-the-box. Could we update it to use Google STUN server instead?

Either way, I think it's worth mentioning in the upcoming wiki page that relying upon default/public STUN servers is risky because they tend to be taken offline, and troubleshooting WebRTC issues should always begin with checking STUN/TURN server availability.

**Comment by dteleguin** [comment]
@popindavibe also thanks for sharing the config. Seems like we've missed the important `relay-ip` option. Without it, coturn's behavior is quite different from what we wanted.

As for exposing TURN credentials, it's a well-known problem that has been addressed more or less. Basically, we have the following options:
* Long-term credentials (`-a, --lt-cred-mech` coturn flags), aka traditional login-password authentication. This means TURN server maintains a user database. This is the simplest variant, but, as you've mentioned, it's insecure in the sense of exposing plain-text credentials even to unauthenticated users;
* Ephemeral credentials aka TURN REST API (`--use-auth-secret`). This is more secure because clients don't have access to any persistent credentials. Instead, they obtain temporary credentials from an URL. The issuing server and the TURN server should share a cryptographic secret so that the credentials could be verified on the TURN side. (The process is described in full detail [here](https://tools.ietf.org/html/draft-uberti-behave-turn-rest-00).)

The problem is, JSXC is a purely client-side solutions and does not (and probably should not) provide any server-side code; so it's your responsibility to provide an URL of an endpoint that would issue credentials, trusted by your TURN server. I think this logic could be implemented as a part of an integration plugin.

BTW, we've revived a [JSXC RoundCube plugin](https://github.com/AcutusPro/jsxc.roundcube) (@sualko, pls see my ML posting on that). I think I'll give it a try and integrate ephemeral credential issuing into the plugin as a proof-of-concept, hopefully by the end of the week.

* oAuth token-based authentication/authorization (I'll write about it a bit later).

**Comment by popindavibe** [comment]
@dteleguin Well, I'm sure I saw google's public stun server in SOGo's plugin, I first saw it through packet analyzing, and found it later on in the JS.
I'm all for ephemeral, I just don't know where to start and configure it on both ends.

**Comment by sualko** [comment]
@dteleguin 

> seems like JSXC ships with non-working STUN config out-of-the-box

It is working for me and `host stun.stunprotocol.org` is returning `stun.stunprotocol.org has address 52.86.10.164` on my machine. Maybe it is blocked by your administrator?

> Could we update it to use Google STUN server instead?

I try to avoid as many google services as possible :wink:.

Yes I saw the ML post about roundcube and I will try to answer it this week. Please remember me, if I forget it once again.

Yes, ephemeral credentials is the way to go. This is the reason why most of my maintained plugins get their webrtc configuration from a php script (e.g. sogo `ajax/getturncredentials.php`). For an example use of the turn rest api you can checkout my owncloud implementation of it: https://github.com/owncloud/jsxc.chat/blob/master/ajax/getTurnCredentials.php

**Comment by popindavibe** [comment]
oh, so here is the why of the php file ;-)
I'm very busy these days, but I'll definitely try to get it work from the implementation example, and will get back here to provide feedback on it.

**Comment by dteleguin** [comment]
I've just added TURN REST API and ephemeral credentials support to the RoundCube plugin. It's implemented in a way that is slightly different (since RC doesn't allow bare PHP pages), but the idea and the algorithm are the same.

@popindavibe, AFAIK you're interested in SOGo first of all. Seems like the SOGo plugin used to implement TURN REST API once (see [https://github.com/jsxc/jsxc.sogo/blob/master/ajax/getturncredentials.php]), but it was later replaced by simple static credentials. I hope @sualko will shed some light on that.

Another option is obtaining STUN/TURN servers (and optionally credentials) from the XMPP server itself. There exists a dedicated [XEP-0215: External Service Discovery](http://xmpp.org/extensions/xep-0215.html).
The good news is, this XEP is already supported by Strophe.jinglejs (and by JSXC as a consequence). The bad news is, ejabberd supports it only in its commercial version, and [support in Prosody](https://hg.prosody.im/prosody-modules/file/tip/mod_extdisco/mod_extdisco.lua) is rudimentary, i.e. it can return only statically configured servers and static (not generated) credentials. However, in the case of Prosody it shouldn't be that hard to properly implement ephemeral credentials; Lua is clear and understandable, contrary to the Secret Alien Technology™ ejabberd is written in :)

**Comment by dteleguin** [comment]
@sualko as for stun.stunprotocol.org, I've done some research, and it seems like it resolves to 127.0.0.1 for Russia and parts of Ukraine (at least). Definitely a GeoIP/GeoDNS issue here; maybe has something to do with this: [DDOS on stun.stunprotocol.org](https://groups.google.com/forum/#!topic/stunprotocol/7b7i6jlAVTs). Setting IP to 127.0.0.1 is a well-known technique to fight DDoS.

I think it's ok to have stun.stunprotocol.org as a default, provided that the docs clearly state that the server might be unreachable from some parts of the world.

**Comment by sualko** [comment]
What do you mean with  "implemented TURN REST API" once? The [php file](https://github.com/jsxc/jsxc.sogo/tree/master/ajax) is still there, but there was never a turn rest api implemented.

@dteleguin Thank you for this very helpful information. Never thought an something like that. Maybe you want to add this note to this brand [new wiki page](https://github.com/jsxc/jsxc/wiki/WebRTC-how-to).

Is there still an issue with webrtc, or can we close this issue?

**Comment by popindavibe** [comment]
@sualko @dteleguin 
I tried putting what I believe should be how ephemeral is supposed to be presented to the turn server, but coturn does not accept my user in the form 'user:timestamp' saying it is not in its database.

Although for now I have no real-life solution (SOGo does not use PHP, I would like to not have to install it just for symmetrical NAT issue...), I wanted to manually validate via following conf (in sjsxc.config.js) : 
```
{
               urls: ['turn:turn.example.com','turn:turn2.example.com'],
               username: 'testwebrtc:1482162550',
               credential: '1mfB852servz6xco=',
               credentialType: 'token'
}
```

In coturn's turnserver.conf I configured (relevant parts) : 
```
fingerprint
lt-cred-mech
use-auth-secret
static-auth-secret
userdb=/var/lib/turn/turndb
realm=example.com
```
I inserted SECRET via turnadmin in sqlite DB
`turnadmin --db /var/lib/turn/turndb -r example.com --set-secret=XXXXXXX`

But everytime I try to connect I get 
```
174: check_stun_auth: Cannot find credentials of user <testwebrtc:1482162550>
174: session 004000000000000001: realm <example.com> user <testwebrtc:1482162550>: incoming packet message processed, err
or 401: Unauthorized
```
I'm out of options...

**Comment by dteleguin** [comment]
@popindavibe,
First, the format is `timestamp:username`, not vice versa. Remember this should be an expiration timestamp, not "now" (usually a "now" + TTL in seconds). You can consult the [code](https://github.com/AcutusPro/jsxc.roundcube/blob/master/jsxc.php#L110) that has proven to be working with coturn.

Also you'd need to remove the `lt-cred-mech` option, otherwise coturn wouldn't switch into the auth secret mode.

**Comment by popindavibe** [comment]
@dteleguin 
I'm using Openfire. I searched for its compatibility with XEP-0215 but info is scarce and Wikipedia compatibility's Array shows it as unknown regarding that extension implementation. 
Thanks for the format of the login, I saw a different info on the Web, I'll try that and remove the lt-cred-mech as well.

**Comment by dteleguin** [comment]
@sualko,

> What do you mean with "implemented TURN REST API" once? The php file is still there, but there was never a turn rest api implemented.

Please disregard that - I've was confused by the [initial version](https://github.com/jsxc/jsxc.sogo/blob/edc5a33bddf0be5ef86f61e0f9383862effa3e0d/ajax/getturncredentials.php) of that file, but now I see that it was simply copied from ownCloud plugin. (AFAIK, SOGO 3 has been completely rewritten in Obj-C.)

> Thank you for this very helpful information. Never thought an something like that. Maybe you want to add this note to this brand new wiki page.

I've added a section on public STUN/TURN servers and their peculiarities.

> Is there still an issue with webrtc, or can we close this issue?

I'm afraid there's still a problem with filetransfer over TURN relay. It only works for Chrome-Chrome transfers + very specific TURN server configuration. No other browser combinations seem to work. The symptoms are always the same: the sending side fails to collect non-local ICE candidates (i.e. collects only `host` type candidates, and neither `srflx` nor `relay`). Thus, no working candidate pair is determined.

I believe this is not an issue of JSXC per se, but is rather related to Jingle.JS. Its last significant update has happened about 1.5 years ago - has it become abandonware? They've even missed an important dependency update (see this [PR](https://github.com/otalk/jingle.js/pull/32)), thus they're using `webrtcsupport` version that is more than two years old. Could you somehow influence that? There have been many changes to WebRTC stacks during last years, so first of all we should make sure that all the adaptors and libraries are up to date.

Unfortunately, I haven't yet managed to test voice/video calls over TURN. It will take some time since I'll need to find two boxes with cameras connected to different networks, so I'll ask you please not to close this issue.

Additionally, I could help you deploy a private TURN server at jsxc.org if you wish, so that the project members could at least reproduce the bugs I'm talking about.

**Comment by popindavibe** [comment]
Ok, so it's working manual-style with or without lt-cred-mech (but there's no point keeping it).
@dteleguin as for video-chat, I could test the turn server on several occasion and it was working fine. The file transfer was not working though (well, whatever the kind of link between the two parties).

@sualko : I would also like to keep it open as I am currently trying to use python or bash to get the hash generated by the webserver. Python is supposed to have modules ready to do that, but I can't get it generate the hash propertly yet, so I'll probably end up try implementing as a shell script as a POC for ephemeral without PHP (which makes a lot of sense for SOGo).

**Comment by dteleguin** [comment]
@popindavibe thanks for the info, it seems like jingle.js uses different IP collection mechanisms for voice/video and file transfers. Quite plausible because they are implemented by different jingle.js modules, namely [jingle-media-session](https://github.com/otalk/jingle-media-session) and [jingle-filetransfer-session](https://github.com/otalk/jingle-filetransfer-session).

BTW, here's a Python snippet for ephemerals ([pycrypto](http://www.pycrypto.org) should be installed):
```python
import time
import Crypto.Hash.SHA
from base64 import b64encode
from Crypto.Hash import HMAC

ttl = 86400
timestamp = int(time.time()) + ttl
username = str(timestamp) + ':username'
secret = 'secret'
hmac = HMAC.new(secret, username, Crypto.Hash.SHA)
password = b64encode(hmac.digest())

print(password)
```

**Comment by popindavibe** [comment]
Ok, all good and tested successfully behind a Symmetrical NAT.
I'll edit SOGo's Wiki here and add a [link to the project](https://gitlab.nomagic.fr/popi/jsxc-rtcpeerconfig)  (files and howto) I created.
As far as I'm concerned, this issue can be closed.

**Comment by sualko** [comment]
@popindavibe Thanks for your wiki entry.

**Comment by diegobill** [comment]
@christophschw , Can you make a video call using turn:numb.viagenie.ca ? My configs does not work in issue #583 .
